apiserver/common/networkingcommon/networkconfigapi.go

```diff
@@ -117,29 +129,229 @@ func (api *NetworkConfigAPI) getMachineForSettingNetworkConfig(machineTag string
+// that are parents of NICs in the incoming set.
+// This *should* be impossible, but we can be defensive without throwing an
+// error - any unobserved devices will always have their addresses removed
+// provided that they are under the authority of the machine.
+// See processRemovalCandidates.
+func (o *updateMachineLinkLayerOp) noteObservedParentDevices() {
+	for _, dev := range o.incoming {
+		if dev.ParentInterfaceName != "" {
+			o.observedParentDevices.Add(dev.ParentInterfaceName)
+		}
+	}
+}
 
-	logger.Debugf("setting devices: %+v", devicesArgs)
-	if err := m.SetParentLinkLayerDevicesBeforeTheirChildren(devicesArgs); err != nil {
-		return errors.Trace(err)
+func (o *updateMachineLinkLayerOp) processExistingDevice(dev LinkLayerDevice) ([]txn.Op, error) {
+	incomingDev := o.MatchingIncoming(dev)
+
+	if incomingDev == nil {
+		ops, err := o.processExistingDeviceNotObserved(dev)
+		return ops, errors.Trace(err)
 	}
 
-	logger.Debugf("setting addresses: %+v", devicesAddrs)
-	if err := m.SetDevicesAddresses(devicesAddrs...); err != nil {
-		return errors.Trace(err)
+	ops := dev.UpdateOps(networkDeviceToStateArgs(*incomingDev))
+
+	incomingAddrs := o.MatchingIncomingAddrs(dev.MACAddress())
+
+	for _, addr := range o.DeviceAddresses(dev) {
+		existingAddrOps, err := o.processExistingDeviceAddress(dev, addr, incomingAddrs)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ops = append(ops, existingAddrOps...)
 	}
 
-	logger.Debugf("updated machine %q network config", m.Id())
-	return nil
+	newAddrOps, err := o.processExistingDeviceNewAddresses(dev, incomingAddrs)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	o.MarkDevProcessed(dev.MACAddress())
+	return append(ops, newAddrOps...), nil
+}
+
+// processExistingDeviceNotObserved returns transaction operations for
+// processing a device we have in state, but that the machine agent no
+// longer observes locally.
+// The device itself is not marked for deletion now, but for later processing
+// to ensure it is not a parent of other observed devices.
+func (o *updateMachineLinkLayerOp) processExistingDeviceNotObserved(dev LinkLayerDevice) ([]txn.Op, error) {
+	addrs := o.DeviceAddresses(dev)
+
+	var ops []txn.Op
+	var removing int
+	for _, addr := range addrs {
+		// If the machine is the authority for this address,
+		// we can delete it; otherwise leave it alone.
+		if addr.Origin() == network.OriginMachine {
+			logger.Debugf("removing address %q from device %q", addr.Value(), dev.Name())
+			ops = append(ops, addr.RemoveOps()...)
+			removing++
+		}
+	}
+
+	// If the device is having all of its addresses removed and is not under
+	// the authority of the provider, add it as a candidate for removal.
+	// If the device has been relinquished by the provider, the instance-poller
+	// will have removed the provider ID - see the instance-poller API facade
+	// logic.
+	if removing == len(addrs) && dev.ProviderID() == "" {
```

Comment:
If I'm reading this correctly we get to keep machiner-managed devices even if they have no addresses left. Is there a particular reason for keeping these while dropping the provider ones?

Reply:
We only remove addresses if the provider is *not* the authority for them.

There *might* be an eventually consistent thing here where the machine agent no longer sees the device, but it is only removed after the provider agrees that it is gone and relinquishes the address origin.

If the addresses are all gone, and the machine is the authority we remove it.